### PR TITLE
switch-root: Make sure all mounts under /run survive switch-root

### LINKED ISSUE #28452: systemd forgets mounts during transition from initrd
### systemd version the issue has been seen with

254-rc2

### Used distribution

Fedora Rawhide

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

We've got a couple of reports [[0]](https://github.com/coreos/fedora-coreos-tracker/issues/1527)[[1]](https://bugzilla.redhat.com/show_bug.cgi?id=2223795) regarding missing mounts and after some playing around with [Fedora-Server-dvd-x86_64-Rawhide-20230718.n.0.iso](https://dl.fedoraproject.org/pub/fedora/linux/development/rawhide/Server/x86_64/iso/Fedora-Server-dvd-x86_64-Rawhide-20230718.n.0.iso) it looks like systemd forgets mounts mounted in initrd when transitioning to the "real" system:

Reproducer:
```
$ qemu-kvm -boot d -cdrom ~/Downloads/Fedora-Server-dvd-x86_64-Rawhide-20230718.n.0.iso -m 2G -nographic

## Tweak the kernel command line in grub and add: rd.break console=ttyS0
sh-5.2# mkdir /run/{foo,bar}
sh-5.2# mount --bind -v /run/foo /run/bar
mount: /run/foo bound on /run/bar.
sh-5.2# mount -l | grep /run
tmpfs on /run type tmpfs (rw,nosuid,nodev,size=400116k,nr_inodes=819200,mode=755,inode64)
/dev/sr0 on /run/install/repo type iso9660 (ro,relatime,nojoliet,check=s,map=n,blocksize=2048,iocharset=utf8) [Fedora-S-dvd-x86_64-rawh]
/dev/loop0 on /run/rootfsbase type squashfs (ro,relatime,errors=continue)
LiveOS_rootfs on /sysroot type overlay (rw,relatime,lowerdir=/run/rootfsbase,upperdir=/run/overlayfs,workdir=/run/ovlwork)
tmpfs on /run/bar type tmpfs (rw,nosuid,nodev,size=400116k,nr_inodes=819200,mode=755,inode64)
sh-5.2# exit

## Switch to the second tmux pane with ^B-2
# mount -l | grep /run
LiveOS_rootfs on / type overlay (rw,relatime,seclabel,lowerdir=/run/rootfsbase,upperdir=/run/overlayfs,workdir=/run/ovlwork)
tmpfs on /run type tmpfs (rw,nosuid,nodev,seclabel,size=400112k,nr_inodes=819200,mode=755,inode64)
```

The same thing with [Fedora-Server-dvd-x86_64-38-1.6.iso](https://download.fedoraproject.org/pub/fedora/linux/releases/38/Server/x86_64/iso/Fedora-Server-dvd-x86_64-38-1.6.iso) works as expected (i.e. the mounts are present after the initrd transition).

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by daandemeyer**
In https://github.com/systemd/systemd/commit/7c764d45997721705e43ac66fbb5a56747d00d40, we changed behavior
from recursively moving all mounts in /run from the initrd to the
rootfs on switch-root to only moving /run itself from the initrd to
the rootfs on switch-root. This breaks applications that rely on
mounts in /run to survive the switch-root operation.

Let's switch back to recursively moving all mounts from /run to the
rootfs to keep backwards compat intact.

### CODE REVIEW

File: src/shared/switch-root.c [hunk 1]
```diff
@@ -37,7 +37,7 @@ int switch_root(const char *new_root,
                 { "/dev",                                 MS_BIND|MS_REC }, /* Recursive, because we want to save the original /dev/shm + /dev/pts and similar */
                 { "/sys",                                 MS_BIND|MS_REC }, /* Similar, we want to retain various API VFS, or the cgroupv1 /sys/fs/cgroup/ tree */
                 { "/proc",                                MS_BIND|MS_REC }, /* Similar */
-                { "/run",                                 MS_BIND        }, /* Stuff mounted below this we don't save, as it might have lost its relevance, i.e. credentials, removable media and such, we rather want that the new boot mounts this fresh */
+                { "/run",                                 MS_BIND|MS_REC }, /* Stuff mounted below this we don't save, as it might have lost its relevance, i.e. credentials, removable media and such, we rather want that the new boot mounts this fresh */
                 { SYSTEM_CREDENTIALS_DIRECTORY,           MS_BIND        }, /* Credentials passed into the system should survive */
                 { ENCRYPTED_SYSTEM_CREDENTIALS_DIRECTORY, MS_BIND        }, /* Similar */
```
Comment (keszybz) [inline src/shared/switch-root.c]: The comment is outdated now.

And it seems iffy to exclude SYSTEM_CREDENTIALS_DIRECTORY and  ENCRYPTED_SYSTEM_CREDENTIALS_DIRECTORY. Normally, we don't expect any mounts underneath those. But if there *are* any, then they should be propagated. Thus, I think this `transfer_table` should be made into a strv, and the same `MS_BIND|MS_REC` flags should be used everywhere.

While at it, please fix the adjective/adverb confusion for "Similar" and "Recursive".

**Review by YHNdnzj — APPROVED** [review]

**Comment by bluca** [comment]
Please cherry pick the reproducer from https://github.com/systemd/systemd/pull/28456 as it doesn't make sense to run the CI twice

**Comment by keszybz** [comment]
The test with the reproducer is in https://github.com/systemd/systemd/pull/28457

**Comment by mrc0mmand** [comment]
> > Spoiler alert - the CIs will start failing in a bit, as this change breaks TEST-82-SOFTREBOOT
> 
> I guess for `soft-reboot` we definitely want to get rid of sub-mounts of `/run` from the previous rootfs. So this needs some conditionalization...

Yeah, this might be a pickle, since the MS_BIND stuff was introduced as part of the soft-reboot PR (https://github.com/systemd/systemd/pull/27435/commits).

Also, reading through the 7c764d45997721705e43ac66fbb5a56747d00d40 commit again, I wonder about this part:

> Traditionally, using MS_MOVE was preferable, because we didn't bother
with unmounting the old mount hierarchy before the switch root, and thus
using MS_MOVE did some clean-up as side-effect (because the old mounts
went away this way). But since we nowadays properly umount all remaining
mount points (since https://github.com/poettering/systemd/commit/268d1244e87a35ff8dff56c92ef375ebf69d462e) when
transitioning it's pointless.

I don't think the second half of that statement is true, since with 268d1244e87a35ff8dff56c92ef375ebf69d462e we do unmount (almost) all mounts via `umount_recursive_full()` but _only_ in the fallback path if `pivot_root()` fails. Or maybe I'm misunderstanding this :shrug:.

Anyway, one possible route could be making use of the `flags` field introduced by 2932161bf6d9a8c127da00c9032e7e449222c9ea to make this work with soft-reboot.

**Comment by bluca** [comment]
> > > Spoiler alert - the CIs will start failing in a bit, as this change breaks TEST-82-SOFTREBOOT
> > 
> > 
> > I guess for `soft-reboot` we definitely want to get rid of sub-mounts of `/run` from the previous rootfs. So this needs some conditionalization...
> 
> Yeah, this might be a pickle, since the MS_BIND stuff was introduced as part of the soft-reboot PR (https://github.com/systemd/systemd/pull/27435/commits).
> 
> Also, reading through the [7c764d4](https://github.com/systemd/systemd/commit/7c764d45997721705e43ac66fbb5a56747d00d40) commit again, I wonder about this part:
> 
> > Traditionally, using MS_MOVE was preferable, because we didn't bother
> > with unmounting the old mount hierarchy before the switch root, and thus
> > using MS_MOVE did some clean-up as side-effect (because the old mounts
> > went away this way). But since we nowadays properly umount all remaining
> > mount points (since [poettering@268d124](https://github.com/poettering/systemd/commit/268d1244e87a35ff8dff56c92ef375ebf69d462e)) when
> > transitioning it's pointless.
> 
> I don't think the second half of that statement is true, since with [268d124](https://github.com/systemd/systemd/commit/268d1244e87a35ff8dff56c92ef375ebf69d462e) we do unmount (almost) all mounts via `umount_recursive_full()` but _only_ in the fallback path if `pivot_root()` fails. Or maybe I'm misunderstanding this shrug.
> 
> Anyway, one possible route could be making use of the `flags` field introduced by [2932161](https://github.com/systemd/systemd/commit/2932161bf6d9a8c127da00c9032e7e449222c9ea) to make this work with soft-reboot.

--> https://github.com/systemd/systemd/pull/28497

**Comment by bluca** [comment]
superseded by https://github.com/systemd/systemd/pull/28497

**Pull request closed without merging**